File: Java/src/main/java/QuestionBank/Solution801_900/Solution860.java
package QuestionBank.Solution801_900;

/**
 * @author windows
 */
public class Solution860 {
    public boolean lemonadeChange(int[] bills) {
        if (bills[0] != 5){
            return false;
        }
        int[] money = new int[3];
        money[0]++;
        for (int i = 1; i < bills.length; i++) {
            if (bills[i] == 5){
                money[0]++;
            }else if(bills[i] == 10){
                if (money[0] < 1){
                    return false;
                }
                money[0]--;
                money[1]++;
            }else{
                if (money[1] == 0){
                    if (money[0] < 3){
                        return false;
                    }else{
                        money[0] -= 3;
                    }
                }else{
                    if (money[0] < 1){
                        return false;
                    }
                    money[0]--;
                    money[1]--;
                }
            }
        }
        return true;
    }


    public static void main(String[] args) {
        Solution860 solution860 = new Solution860();
    }
}

File: C++/src/Solution801_900/Solution860.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution860.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <unordered_map>

bool Solution::lemonadeChange(vector<int> &bills) {
    unordered_map<int, int> map{{5, 0}, {10, 0}, {20, 0}};
    int n = bills.size();
    for (int i = 0; i < n; ++i) {
        if (bills[i] == 10){
            if (--map[5] < 0){
                return false;
            }
            map[10]++;
        }else if (bills[i] == 20){
            if (map[10] < 1){
                if (map[5] < 2){
                    return false;
                }
                map[5] -= 2;
            }else {
                if (map[5] < 1){
                    return false;
                }
                map[10]--;
                map[5]--;
            }
            map[20]++;
        } else{
            map[5]++;
        }
    }
    return true;
}
